src/utils/chart.ts

```diff
@@ -125,3 +127,73 @@ export const getFeeRateSamples = (feeRates: Array<FeeRateTracker.TransactionFeeR
 
   return samples
 }
+
+type AssertIsArray = (
+  value: EChartOption.Tooltip.Format | EChartOption.Tooltip.Format[],
+) => asserts value is EChartOption.Tooltip.Format[]
+
+export const assertIsArray: AssertIsArray = (value: EChartOption.Tooltip.Format | EChartOption.Tooltip.Format[]) => {
+  const isArray = Array.isArray(value)
+  if (!isArray) {
+    throw new Error('value is not an array')
+  }
+}
+
+type AssertIsNotArray = (
+  value: EChartOption.Tooltip.Format | EChartOption.Tooltip.Format[],
+) => asserts value is EChartOption.Tooltip.Format
+
+export const assertNotArray: AssertIsNotArray = (
+  value: EChartOption.Tooltip.Format | EChartOption.Tooltip.Format[],
+) => {
+  const isArray = Array.isArray(value)
+  if (isArray) {
+    throw new Error('value is an array')
+  }
+}
+
+type AssertSerialsItem = (value: EChartOption.Tooltip.Format) => asserts value is SeriesItem
+export const assertSerialsItem: AssertSerialsItem = (value: EChartOption.Tooltip.Format) => {
+  if (
+    typeof value.seriesName !== 'string' ||
+    typeof value.name !== 'string' ||
+    typeof value.color !== 'string' ||
+    typeof value.dataIndex !== 'number'
+  ) {
+    throw new Error('invalid SeriesItem')
+  }
+}
+
+export const assertSerialsDataIsString: (value: EChartOption.Tooltip.Format) => asserts value is { data: string } = (
+  value: EChartOption.Tooltip.Format,
+) => {
+  if (typeof value.data !== 'string') {
+    throw new Error('data is not string')
+  }
+}
+
+export const assertSerialsDataIsStringArrayOf3: (
+  value: EChartOption.Tooltip.Format,
+) => asserts value is { data: [string, string, string] } = (value: EChartOption.Tooltip.Format) => {
+  if (
+    Array.isArray(value.data) ||
+    typeof value.data[0] !== 'string' ||
+    typeof value.data[1] !== 'string' ||
+    typeof value.data[2] !== 'string'
+  ) {
+    throw new Error('invalid SeriesItem length of 3')
+  }
+}
+export const assertSerialsDataIsStringArrayOf4: (
+  value: EChartOption.Tooltip.Format,
+) => asserts value is { data: [string, string, string, string] } = (value: EChartOption.Tooltip.Format) => {
+  if (
+    Array.isArray(value.data) ||
+    typeof value.data[0] !== 'string' ||
+    typeof value.data[1] !== 'string' ||
+    typeof value.data[2] !== 'string' ||
+    typeof value.data[3] !== 'string'
+  ) {
+    throw new Error('invalid SeriesItem length of 4')
+  }
+}
```

Comment:
Why to define the `type` for the first three functions, but not for the last three one?

Reply:
They all have type, but fisrt three are too long, so I didn't inline them:

<img width="800" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/3870972/825b1e25-5fab-4771-81ad-55594817fea9">

we can extract the hilighted part to a type also, if needed